CRM/Core/Payment/OmnipayMultiProcessor.php

```diff
@@ -830,7 +836,10 @@ protected function redirectOrExit($outcome, $response) {
       case 'success':
         $userMsg = NULL;
         $redirectUrl = $this->getStoredUrl('success');
-        if (!$redirectUrl && method_exists($response, 'confirm')) {
+        if (empty($redirectUrl)) {
+          $redirectUrl = $this->getReturnSuccessUrl(CRM_Utils_Request::retrieve('qfKey', 'String'));
+        }
+        if (method_exists($response, 'confirm')) {
```

Comment:
@eileenmcnaughton For sagepay we don't seem to have a stored URL (because we are in the IPN handler?) so we need to generate one here with the session key that is passed back to us via the IPN request URL.